[REFACTOR] OrderService 수정 #176

oneokiwa opened the pull request:
## #️⃣연관된 이슈

<!-- resolves: #이슈번호, #이슈번호 -->
resolves: #176, #177 

## 📝주요 작업 내용
## 1. N+1 문제 해결
<!-- 이번 PR에서 작업한 내용을 간략히 설명해주세요(이미지 첨부 가능) -->
주문 조회(getOrderProducts) 및 관리자 상세 조회(getOrderDetail) 기능에서 발생하던 Order → OrderProduct → Product Lazy 로딩 기반의 N+1 문제를 해결했습니다.
조회 성능을 개선하기 위해 QueryDSL 기반 Fetch Join 쿼리를 적용했습니다.

- 기존에 사용하던 findAllByOrderId()는 조회 시 N+1 위험이 있어
내부 로직에서만 사용했습니다.

## 2. OrderProductStatus로 대체
- 기존 OrderStatus는 삭제하고 모두 OrderProductStatus를 사용하도록 리팩토링.
- 관리자 조회용 페이지에서는 편의를 위해 주문상품들의 상태를 계산한 OrderDerivedStatus를 보여주도록 수정.
- WAITING_PATMENT, PAID 상태 삭제. PAID 상태는 SHIPPING_READY로 대체함.
## 3. Cancel API 수정
- 주문 상태가 없어짐에 따라 주문 단위 취소가 아닌, 주문 상품별 취소로 변경.
- 엔드포인트 주소도 기능에 맞게 수정.

### TODO (2차 스프린트)
- 추후 ShippingDetail 로직이 구현되면 동일한 방식의 Fetch Join 적용이 필요.
- 기존 주문 상태 변경 관련 코드(updateStatus) 삭제후, 새로운 기능으로 대체 필요.
  - 관리자 화면에서 상태를 수동 변경 (ex. 배송 끝났는데 시스템 장애로 SHIPPING에서 멈춤) forceChangeOrderProductStatus() 메서드 추가.
<!-- 리뷰어가 특별히 봐주었으면 하는 부분이 있다면 작성해주세요
ex) 메서드 XXX의 이름을 더 잘 짓고 싶은데 혹시 좋은 명칭이 있을까요? -->

oneokiwa committed:
feat: OrderProduct 상태 추가 (PAID)

oneokiwa committed:
refactor: createOrder 수정
- createOrder 내부에서 재고 차감하는 코드 삭제
- checkStock, reduceStock 메서드 분리

oneokiwa committed:
refactor: order 패키지 구조 수정

oneokiwa committed:
refactor: N+1 문제 해결

tryterry77 commented on src/main/java/com/kt/service/OrderServiceImpl.java:
```diff
@@ -47,12 +47,24 @@ public class OrderServiceImpl implements OrderService {
+		List<OrderProductEntity> orderProducts = orderProductRepository.findWithProductByOrderId(orderId);
 		return OrderResponse.OrderProducts.of(orderId, orderProducts);
 	}
 
+	// TODO: @Lock 붙이기
+	@Transactional(readOnly = true)
+	public void checkStock(List<OrderRequest.Item> items) {
+		for (OrderRequest.Item item : items) {
+			ProductEntity product = productRepository.findByIdOrThrow(item.productId());
+
+			if (product.getStock() < item.quantity()) {
+				throw new CustomException(ErrorCode.STOCK_NOT_ENOUGH);
+			}
+		}
+	}
+
```
checkStock은 아직 개발 중인 코드인가요? 아직 다른 곳에서 사용하고 있지 않네요

tryterry77 commented on src/main/java/com/kt/service/OrderServiceImpl.java:
```diff
@@ -94,6 +100,25 @@ public void createOrder(String email, List<OrderRequest.Item> items, UUID addres
+	public void reduceStock(UUID orderId) {
+
+		OrderEntity order = orderRepository.findByIdOrThrow(orderId);
+		
+		List<OrderProductEntity> orderProducts =
+			orderProductRepository.findAllByOrderId(orderId);
+		
+		for (OrderProductEntity orderProduct : orderProducts) {
+			ProductEntity product = orderProduct.getProduct();
+			Long quantity = orderProduct.getQuantity();
+
+			product.decreaseStock(quantity);
+			orderProduct.updateStatus(OrderProductStatus.PAID);
+		}
```
여기도 마찬가지 입니다.

oneokiwa replied:
네 둘다 리팩토링 끝내고 마저 구현할 예정입니다

junsang-ee commented on src/main/java/com/kt/constant/OrderProductStatus.java:
```diff
@@ -5,7 +5,8 @@
 	CREATED("주문생성"),
-	WAITING_PAYMENT("결제대기"),
+	WAITING_PAYMENT("결제대기"), // TODO: 삭제 필요
+	PAID("결제완료"),
```
생각을 좀 해봤는데 결제완료 라는 문구보다는 배송대기가 좋아보이는데 (결제가 완료 돼야 베송을 할 수 있음 뭐 이런 뉘앙스) 어때보이십니까

oneokiwa replied:
좋아요 배송대기로 수정하겠습니다!

junsang-ee commented on src/main/java/com/kt/service/OrderServiceImpl.java:
```diff
@@ -47,12 +47,24 @@ public class OrderServiceImpl implements OrderService {
 
+	// TODO: @Lock 붙이기
+	@Transactional(readOnly = true)
+	public void checkStock(List<OrderRequest.Item> items) {
```
이 기능 생각을 좀 해봤는데요 

도메인별 서비스를 혼합되게 할필요없을것같구요 오케스트레이션서비스 만들어서 해야할듯합니다

junsang-ee replied:
수정요 오케스트레이션 -> 유스케이스 or 파서드

oneokiwa committed:
refactor: PAID 상태 문구 변경, WAITING_PAYMENT 상태 삭제

oneokiwa committed:
refactor: OrderService 리팩토링
[OrderService]
- cancelOrderProduct: 주문 상품 단위 취소로 변경 (주문 단위 취소 불가능)
- 배송지 정보 수정 기능이므로 changeOrderAddress로 명칭 변경, 하나라도 SHIPPING 상태인 경우 수정 불가능 처리.
- 관리자 전용 기능 삭제 (주문 상태 강제 변경)

oneokiwa committed:
feat: OrderDerivedStatus 추가
[OrderEntity]
- 관리자 조회 페이지용 주문 상태 계산 메서드 추가 (단순 표시용)

oneokiwa committed:
refactor: OrderProductStatus로 대체

oneokiwa committed:
refactor: Admin 주문 상세 조회 수정
- 주문자명 -> 주문자Id로 대체
- OrderStatus 삭제, OrderDerivedStatus로 대체
- 정책 수정이 필요한 부분 주석 처리

oneokiwa committed:
refactor: 테스트 코드 OrderProductStatus로 대체
- AdminOrderServiceTest로 변경 필요한 테스트 Todo 처리

oneokiwa committed:
refactor: 주문 취소 API 엔드포인트 변경 (member, admin)
- /order-products/{orderProductId}/cancel
- 주문상품 단위로 취소 가능

junsang-ee commented on src/main/java/com/kt/constant/OrderProductStatus.java:
```diff
@@ -5,7 +5,7 @@
 public enum OrderProductStatus {
 	CREATED("주문생성"),
-	WAITING_PAYMENT("결제대기"),
+	PAID("배송대기"),
```
PAID 말고 배송대기에 적합한 단어 없을까요? 저도 생각해볼게요

oneokiwa replied:
말그대로 WAITING_SHIPPING으로 바꿀까요

junsang-ee replied:
SHIPPING_PENDING 도 괜찮을듯한데 그리고 이미 SHIPPING_READY 가 있는데 이건 어떻게 할까요

junsang-ee replied:
PAID 결제 완료 -> 배송대기 로 바꾸고
결제실패 -> 주문실패 (FAILED) 로 바꾸죱

그리고 뭔가 OrderStatus 는 없애긴했는데, OrderProductStatus 보다는 OrderStatus 가 더 의미가 명확하지않나요

주문상품상태, 주문상태 이느낌으로만 보면요 어떤가욤

oneokiwa replied:
PAID 상태를 SHIPPING_READY 상태로 합쳐버리면 되지 않을까요??
OrderStatus로 바꾸면 의미가 헷갈릴 것 같아요 주문상품 단위니까 OrderProductStatus로 유지하는 게 나을 것 같습니다!

junsang-ee replied:
좋아요 OrderProductStatus 유지하시고 PAID 상태 지워버리고 SHIPPING_READY 기존꺼로 가죱

junsang-ee commented:
주문 + 결제는 다 머지하시면 제가 바꿔놓겠습니다

그리고 슬랙에 내용 공유해드릴게요~

oneokiwa committed:
fix: 연관관계 수정
- OrderProduct 생성 시 Order ↔ OrderProduct 양방향 연관관계 누락 수정

oneokiwa committed:
fix: 테스트 코드 오류 수정

l-lyun reviewed (approved):
여러 복잡한 조건들 해결하느라 고생하셨습니다!

oneokiwa committed:
fix: AdminOrderService - OrderProductStatus로 수정
- OrderStatus 삭제 완료
- 어드민 정책 수정 필요: TODO 처리

oneokiwa committed:
fix: 기존 PAID 상태 삭제 및 SHIPPING_READY 상태로 통합

oneokiwa committed:
fix: import 경로 수정

oneokiwa committed:
fix: import 경로 수정
